### PR TITLE
Add ERC721Enumerable

### PR DESCRIPTION
**Pull request by immrsd**
Fixes [#1131](https://github.com/OpenZeppelin/cairo-contracts/issues/1131).

Test target has been compiled with no issues.

### CODE REVIEW

File: packages/core-cairo/src/erc721.test.ts [hunk 1]
```diff
@@ -5,6 +5,14 @@ import { printContract } from './print';
 
 import { erc721 } from '.';
 
+const allFeaturesON: Partial<ERC721Options> = {
+  mintable: true,
+  burnable: true,
+  enumerable: true,
+  pausable: true,
```
Comment (ericglau) [inline packages/core-cairo/src/erc721.test.ts:12]: Minor suggestion for consistency:
```suggestion
  pausable: true,
  enumerable: true,
```

**Review by ericglau — COMMENTED** [review]
Just a few comments.  Can you also add a changelog entry in an `Unreleased` section of `packages/core-cairo/CHANGELOG.md`?

**Comment by immrsd** [comment]
> Looks good, thanks! Just a few comments. Can you also add a changelog entry in an `Unreleased` section of `packages/core-cairo/CHANGELOG.md`?

Done!

**Review by ericglau — APPROVED** [review]
LGTM, thanks!

File: packages/core-cairo/src/erc721.test.ts.md [hunk 2]
```diff
@@ -494,6 +494,236 @@ Generated by [AVA](https://avajs.dev).
...
+            ) {␊
+                let mut contract_state = self.get_contract_mut();␊
+                contract_state.erc721_enumerable.before_update(to, token_id);␊
+            }␊
+    ␊
+            fn after_update(␊
+                ref self: ERC721Component::ComponentState<ContractState>,␊
+                to: ContractAddress,␊
+                token_id: u256,␊
+                auth: ContractAddress,␊
```
Comment (andrew-fleming) [inline packages/core-cairo/src/erc721.test.ts.md:584]: We can omit unused hooks (`after_update` in this case). It's not a necessity, just a nice thing to help tidy up the contracts. Definitely outside the scope of the PR

**Review by andrew-fleming — APPROVED** [review]
Nice work, Alejandro! LGTM

File: packages/core-cairo/src/erc721.test.ts.md (hunk 2 above)
Comment (immrsd) [reply packages/core-cairo/src/erc721.test.ts.md:584]: Nice idea, created an issue for that [w#1162](https://github.com/OpenZeppelin/cairo-contracts/issues/1162)

**Comment by ericnordelo** [comment]
Hey @immrsd, the PR looks good. I want to confirm that you did run `yarn run update_scarb_project` to update the `test_project/`, and build with `scarb build` to ensure there are not warnings or compilation errors with the new addition. Let me know if that's the case.

File: packages/core-cairo/src/erc721.ts [hunk 3]
```diff
@@ -75,48 +75,65 @@ export function buildERC721(opts: ERC721Options): Contract {
...
+    c.addFunction(ERC721HooksTrait, {
+      name: 'after_update',
+      args: [
+        { name: 'ref self', type: `ERC721Component::ComponentState<ContractState>` },
+        { name: 'to', type: 'ContractAddress' },
+        { name: 'token_id', type: 'u256' },
+        { name: 'auth', type: 'ContractAddress' },
+      ],
+      code: [],
+    });
```
Comment (ericnordelo) [inline packages/core-cairo/src/erc721.ts:133]: I know we opened an issue to remove this in the future, but for this particular PR I think we can just remove this in the meantime.
```suggestion
```

**Comment by immrsd** [comment]
> Hey @immrsd, the PR looks good. I want to confirm that you did run `yarn run update_scarb_project` to update the `test_project/`, and build with `scarb build` to ensure there are not warnings or compilation errors with the new addition. Let me know if that's the case.

Yes, I executed the script and built with `scarb --release build`, no compilation issues

image

**Pull request merged**